gflownet.py

```diff
@@ -576,49 +647,41 @@ def sample_batch(
                             tl([n_actions]),
                         ]
                     )
+                    # Backward sampling
                     env, state, action, parents, parents_a = self.backward_sample(
-                        env, policy="uniform"
+                        env,
+                        policy="model",
+                        model=model,
+                        temperature=self.temperature_logits,
                     )
                     n_actions += 1
             envs = envs[n_empirical:]
-        # Rest of batch
+        # Policy trajectories
         while envs:
-            states = [env.state2obs() for env in envs]
-            mask = [env.no_eos_mask() for env in envs]
-            random_action = self.rng.uniform()
-            if train is False or random_action > self.random_action_prob:
-                with torch.no_grad():
-                    t0_a_model = time.time()
-                    action_probs = model(tf(states))
-                    if self.mask_eos:
-                        action_probs[mask, -1] = -1000
-                    t1_a_model = time.time()
-                    times["actions_model"] += t1_a_model - t0_a_model
-                    if all(torch.isfinite(action_probs).flatten()):
-                        actions = Categorical(logits=action_probs).sample()
-                    else:
-                        random_action = -1
-                        if self.debug:
-                            print("Action could not be sampled from model!")
-            # TODO: review: random actions sampled for all envs!
-            # TODO: review: if random from numpy, tl(x) gives weird results for scalars
-            if train and random_action < self.random_action_prob:
-                high = (len(self.env.action_space) + 1) * np.ones(len(envs), dtype=int)
-                if self.mask_eos:
-                    high[mask] -= 1
-                actions = self.rng.integers(low=0, high=high, size=len(envs))
+            # Forward sampling
+            if train is False:
+                envs, actions, valids = self.forward_sample(
+                    envs, times, policy="model", model=model, temperature=1.0
+                )
+            else:
```

Comment:
In the training phase, we  could separate the envs in the envs that correspond to policy = "uniform" and those that correspond to policy = "model"!  For each envs to sample an action for, we can generate a random number to chose if it will chose an action according to the model or uniform and we call separately self.forward_sample with a different policy for the two cases.

Reply:
Yes, I had thought of that indeed. The way I would implement it is actually not by separating the envs outside `forward_sample()` but as a new `policy`. Then, inside `forward_sample()` different envs would be sampled differently. However, I believe in practice we will just use `policy="model"` with a `temperature > 1`, since that seems a lot better than uniform sampling.